### ISSUE TITLE
Double clicking CJK text does not select the word under the cursor

### ISSUE THREAD

**Issue by timothyqiu** [godotengine/godot#75808]
### Godot version

4.0.2.stable

### System information

Arch Linux

### Issue description

Double clicking on text selects the word under the cursor.

Try double clicking words below in the browser to see the expected behavior:

```gdscript
func _ready() -> void:
	var variable := "test string"
	var عامل := "سلسلة الاختبار"
	var מִשְׁתַנֶה := "מחרוזת בדיקה"
	var 变量 := "测试 字符串"
	var 変数 := "テスト 文字列"
	var 변수 := "테스트 문자열"
```

This is the actual behavior in the script editor, doubling clicking on CJK text does nothing:

https://user-images.githubusercontent.com/372476/230700739-e82c7f18-baa0-462d-a08a-414c9a66cd30.mp4

It's true that Chinese and Japanese don't use spaces to separate words. Word breaking `字符串` (string) into `字符` (character) and `串` (sequence) is a completely different area of expertise. I think it should at least use spaces as word boundaries.

### Steps to reproduce

Use the test script above.

### Minimal reproduction project

N/A

**Comment by 451507056** [comment]
https://user-images.githubusercontent.com/6943500/234928236-c1b3908b-8f23-49c9-af8a-cfb3968688a5.mp4

https://user-images.githubusercontent.com/6943500/234928288-505bc656-d8b5-4d8f-a934-2bfee9e6316c.mp4

Chinese comments can no longer be double-clicked, which was OK in 3.5.1.

**Comment by Calinou** [comment]
cc @bruvzg

**Comment by Crashim03** [comment]
Hi! I'm a computer science student from Portugal currently completing a course where I have to choose an open source project, choose a bug, and then fix it with a pull request.

I have been looking into this and found out that in the function `shaped_text_get_word_breaks` (from the file text_server.cpp), `l_gl` seems to have duplicated words, i. e. `l_gl[i].start` and `l_gl[i].end` are the same as `l_gl[i + 1].start` and `l_gl[i + 1].end`, and `l_gl[i + 1].flags` are `GRAPHEME_IS_VIRTUAL`, `GRAPHEME_IS_SPACE` and `GRAPHEME_IS_BREAK_SOFT`, instead of just `GRAPHEME_IS_VALID`. Here's a video showing it:

https://github.com/godotengine/godot/assets/93451864/9471a4af-32db-4445-bb4f-c3b41f275329

So the root of the issue is probably in the function `shaped_text_sort_logical`
I will continue to look into this

**Comment by Crashim03** [comment]
I just noticed that this issue is from one year ago 😅. Hope it's still relevant @Calinou @timothyqiu

**Comment by Crashim03** [comment]
Turns out UBreakIterator seems to have a known issue with cjk characters.
Making an exception in `_shaped_text_update_breaks` in text_server_adv.cpp fixes the issue.

```
while (ubrk_next(bi) != UBRK_DONE) {
	int pos = _convert_pos(sd, ubrk_current(bi)) + r_start;
	int pos_p = pos - 1 - sd->start;
	char32_t c = sd->text[pos_p];

	if (is_cjk_character(c)) {
		if (is_cjk_punctuation(c)) {
			sd->breaks[pos] = true;
		}
	}
	else if ((ubrk_getRuleStatus(bi) >= UBRK_LINE_HARD) && (ubrk_getRuleStatus(bi) < UBRK_LINE_HARD_LIMIT)) {
		sd->breaks[pos] = true;
	} else if ((ubrk_getRuleStatus(bi) >= UBRK_LINE_SOFT) && (ubrk_getRuleStatus(bi) < UBRK_LINE_SOFT_LIMIT)) {
		sd->breaks[pos] = false;
	}
	if (pos - sd->start != sd->end && !is_whitespace(c) && (c != 0xfffc)) {
		sd->break_inserts++;
	}
}
```

I also added the following to char_utils.h

```
static _FORCE_INLINE_ bool is_cjk_character(char32_t c) {
	return (c >= 0x4E00 && c <= 0x9FFF) ||(c >= 0x3400 && c <= 0x4DBF) || (c >= 0x3040 && c <= 0x30FF) || (c >= 0x31F0 && c <= 0x31FF) || (c >= 0xAC00 && c <= 0xD7AF) || (c >= 0x1100 && c <= 0x11FF) || (c >= 0x3130 && c <= 0x318F);
}

static _FORCE_INLINE_ bool is_cjk_punctuation(char32_t c) {
    return (c >= 0x3000 && c <= 0x303F) || (c >= 0xFF00 && c <= 0xFFEF);
}
```
I will create a unit test and then a pull request 👍

**Comment by akien-mga** [comment]
Fixed by #90732.
